### PR TITLE
[Merged by Bors] - Deploy default and support custom affinities

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

For https://github.com/stackabletech/issues/issues/323

## Review Checklist

- [ ] Code contains useful comments
- [ ] CRD change approved (or not applicable)
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by sbernauer** [comment]
Started test https://ci.stackable.tech/job/trino-operator-it-custom/57/

File: rust/crd/src/affinity.rs [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+                    preferred_during_scheduling_ignored_during_execution: Some(vec![
+                        affinity_between_role_pods(APP_NAME, cluster_name, &role.to_string(), 70),
+                    ]),
+                    required_during_scheduling_ignored_during_execution: None,
+                }),
+                node_affinity: None,
+                node_selector: None,
+            }
+        }
+    }
```
Comment (razvan) [inline rust/crd/src/affinity.rs:93]: Less duplication
```suggestion
{
    let affinity_between_cluster_pods = affinity_between_cluster_pods(APP_NAME, cluster_name, 20);
    let mut affinities = vec![affinity_between_cluster_pods];
    let additional_affinities: Vec<WeightedPodAffinityTerm> = match role {
        TrinoRole::Coordinator => {
            trino_catalogs
                .iter()
                .filter_map(|catalog| match &catalog.spec.connector {
                    TrinoCatalogConnector::Hive(hive) => Some(&hive.metastore.config_map),
                    TrinoCatalogConnector::Iceberg(iceberg) => Some(&iceberg.metastore.config_map),
                    TrinoCatalogConnector::BlackHole(_)
                    | TrinoCatalogConnector::GoogleSheet(_)
                    | TrinoCatalogConnector::Tpcds(_)
                    | TrinoCatalogConnector::Tpch(_) => None,
                })
                .map(|hive_cluster_name| {
                    affinity_between_role_pods(
                        "hive",
                        hive_cluster_name, // The discovery cm has the same name as the HiveCluster itself
                        "metastore",
                        50,
                    )
                })
                .collect()
        }
        TrinoRole::Worker => {
            trino_catalogs
                .iter()
                .filter_map(|catalog| match &catalog.spec.connector {
                    TrinoCatalogConnector::Hive(hive) => {
                        hive.hdfs.as_ref().map(|hdfs| &hdfs.config_map)
                    }
                    TrinoCatalogConnector::Iceberg(iceberg) => {
                        iceberg.hdfs.as_ref().map(|hdfs| &hdfs.config_map)
                    }
                    TrinoCatalogConnector::BlackHole(_)
                    | TrinoCatalogConnector::GoogleSheet(_)
                    | TrinoCatalogConnector::Tpcds(_)
                    | TrinoCatalogConnector::Tpch(_) => None,
                })
                .map(|hdfs_cluster_name| {
                    affinity_between_role_pods(
                        "hdfs",
                        hdfs_cluster_name, // The discovery cm has the same name as the HdfsCluster itself
                        "datanode",
                        50,
                    )
                })
                .collect()
        }
    };
    affinities.extend(additional_affinities);
    StackableAffinityFragment {
        pod_affinity: Some(PodAffinity {
            preferred_during_scheduling_ignored_during_execution: Some(affinities),
            required_during_scheduling_ignored_during_execution: None,
        }),
        pod_anti_affinity: Some(PodAntiAffinity {
            preferred_during_scheduling_ignored_during_execution: Some(vec![
                affinity_between_role_pods(APP_NAME, cluster_name, &role.to_string(), 70),
            ]),
            required_during_scheduling_ignored_during_execution: None,
        }),
        node_affinity: None,
        node_selector: None,
    }
}
```

**Review by razvan — APPROVED** [review]
lgtm. one suggestion

**Comment by sbernauer** [comment]
Thx!

**Pull request closed without merging**